src/coreclr/jit/lowerarmarch.cpp

```diff
@@ -1208,10 +1208,88 @@ GenTree* Lowering::LowerHWIntrinsic(GenTreeHWIntrinsic* node)
             LowerHWIntrinsicFusedMultiplyAddScalar(node);
             break;
 
+        case NI_Sve_ConditionalSelect:
+        {
+            GenTree* op1 = node->Op(1);
+            if (op1->OperIsHWIntrinsic(NI_Sve_ConvertVectorToMask))
+            {
+                // For cases where conditional-select is used in the user code explicitly,
+                // select one of the true/false value, see if the `cond == 0` and if yes,
+                // just pick the `falseValue` and if `cond == 1`, pick the `trueValue`
+                // without having to do the condition.
+
+                assert(varTypeIsMask(op1));
+
+                GenTreeHWIntrinsic* vectorToMask = op1->AsHWIntrinsic();
+                GenTree*            maskValue    = vectorToMask->Op(2);
+
+                GenTree* op2    = node->Op(2);
+                GenTree* op3    = node->Op(3);
+                GenTree* result = nullptr;
+                if (maskValue->IsVectorZero())
+                {
+                    result = op3;
+                }
+                else if (maskValue->IsVectorAllBitsSet())
+                {
+                    result = op2;
+                }
+
+                if (result != nullptr)
+                {
+                    result = comp->gtCloneExpr(result);
```

Comment:
I'm not sure I see the need/reason for the clone here and I'm pretty sure `clone` itself is incorrect for `lowering` regardless.

Nodes have 1 user and we're removing `node` so regardless of whether `op2` or `op3` is the result, they will no longer have a `use` and so we can simply `use.ReplaceWith(result)` directly without the need to clone

-- If we did need to clone (which as per the above I'm pretty sure we don't actually need to do), then we don't want `gtCloneExpr`. We instead want `ReplaceWithLclVar` to ensure all node kinds are correctly handled.

I then think we want to explicitly call `SetUnusedValue()` on the other operand and `maskValue`, or even simply remove them if they have no side effects.

Notably this also seems like the kind of optimization we'd ideally do much earlier as part of general constant folding, that way everything else later on is much cheaper/simpler.

Reply:
moved it to importer.